### ISSUE TITLE
AssertionError start <= cf < start + freq

### ISSUE THREAD

**Issue by dsevero** [bits-back/bits-back#6]
Hi, @j-towns. I've been using your rANS code for a quite a bit now (thanks again, great stuff). However, I **sometimes** run into the following error

```
Traceback (most recent call last):
  File "util.py", line 66, in pop
    state, symbol = rans.pop_symbol(statfun, precision)(state)
  File "rans.py", line 98, in pop_
    return pop(msg, statfun, precision)
  File "rans.py", line 80, in pop
    symb, (start, prob) = statfun(cf)
  File "util.py", line 49, in dec
    assert start <= cf < start + freq
AssertionError
```

It is a bit odd, because it doesn't always happen even when given the same configuration of runs. Any hints on what it might be?

Thanks

**Comment by j-towns** [comment]
Hey Daniel, sorry it's taken me so long to reply to this. At some point it would be good to improve our error messages to explain the cause. In the case of this assert, let's look at [the context](https://github.com/bits-back/bits-back/blob/094aa7f2625eff49f1fc9b2635f4851bf0943c82/util.py#L45-L51):
```python
def non_uniform_dec_statfun(ppf, cdf):
    def dec(cf):
        idx = ppf(cf)
        start, freq = non_uniform_enc_statfun(cdf)(idx)
        assert start <= cf < start + freq
        return idx, (start, freq)
    return dec
```
The job of `dec` is to take an integer in the range [0, 2 ** prec], which is `cf`, map it to a symbol `idx` and also the pair `(start, freq)` which specify the interval, containing `cf`, which corresponds to this sample. The assert is simply checking that `cf` is indeed in the interval which it's supposed to be in.

If this assert is failing sometimes, that suggests that the ppf and cdf don't always satisfy the pseudo-inverse property which they're supposed to, the property is written out in a [comment lower down](https://github.com/bits-back/bits-back/blob/master/util.py#L82-L84):
```python
# For decodability we must satisfy
#     all(ppf(cf) == s for s in range(1 << prior_prec) for cf in
#         range(cdf(s), cdf(s + 1)))
```
So basically, if you're writing your own cdf and ppf functions, you should test them by asserting that the condition in that comment is always satisfied. We do that in util_test with [`check_cdf_inverse`](https://github.com/bits-back/bits-back/blob/094aa7f2625eff49f1fc9b2635f4851bf0943c82/util_test.py#L13-L18) for the cdf/ppf pairs there, but I suspect we don't perfectly cover all possible cases and I wouldn't be surprised if there's a bug in one of our cdf/ppf pairs causing this issue.

Does that make sense? Are you using your own cdf/ppf or one of the pairs defined in util.py?

**Comment by dsevero** [comment]
Hey! Sorry for the delay.

That makes sense. I was using the gaussian and uniform cdf/ppf functions, nothing fancy. Didn't write my own.

It isn't an issue anymore, because I'm actually using Craystack now. Probably one of the best bleeding-edge-research-oriented package designs I've seen in a long time. I love how it strayed away from the object oriented hype in favor of a simpler functional + nametuple design.

Congrats and thanks!
